### PR TITLE
Dash Routing: Direct, Vnet_direct and Drop action support

### PR DESCRIPTION
**Pull request by mukeshmv**
Support for remaining Phase 1 routing actions as specified in the Sonic HLD / Vnet-Vnet doc.
Fixes issue #134
It generates following additional SAI attributes.

```
/**
 * @brief Attribute data for #SAI_OUTBOUND_ROUTING_ENTRY_ATTR_ACTION
 */
typedef enum _sai_outbound_routing_entry_action_t
{
    SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_VNET,

    SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_VNET_DIRECT,    

    SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_DIRECT,

    SAI_OUTBOUND_ROUTING_ENTRY_ACTION_DROP,

} sai_outbound_routing_entry_action_t;

typedef enum _sai_outbound_routing_entry_attr_t
{
    SAI_OUTBOUND_ROUTING_ENTRY_ATTR_START,

    SAI_OUTBOUND_ROUTING_ENTRY_ATTR_ACTION = SAI_OUTBOUND_ROUTING_ENTRY_ATTR_START,

     /**
     * @brief Action route_vnet, route_vnet_direct parameter DEST_VNET_VNI
     *
     * @type sai_uint32_t
     * @flags CREATE_AND_SET
     * @condition SAI_OUTBOUND_ROUTING_ENTRY_ATTR_ACTION == SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_VNET or SAI_OUTBOUND_ROUTING_ENTRY_ATTR_ACTION == SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_VNET_DIRECT
     * @default 0
     */ 
     SAI_OUTBOUND_ROUTING_ENTRY_ATTR_DEST_VNET_VNI,

    SAI_OUTBOUND_ROUTING_ENTRY_ATTR_IS_OVERLAY_NH_IP_V6,

    SAI_OUTBOUND_ROUTING_ENTRY_ATTR_OVERLAY_NH_IP,

    SAI_OUTBOUND_ROUTING_ENTRY_ATTR_COUNTER_ID,

}

```

### CODE REVIEW

File: dash-pipeline/bmv2/dash_outbound.p4 [hunk 1]
```diff
@@ -24,8 +24,30 @@ control outbound(inout headers_t hdr,
...
 
-    action route_vnet(bit<24> dest_vnet_vni) {
+    /* TODO:
+       If needed we can have a separate route_vnet_direct action for the
+       overlay_nh_ip override. But for this we need to fix the sai_api_gen tool
+       limitation of unique action param names across different actions in the
+       same table. Since the dest_vnet_vni param will be common for route_vnet
+       and route_vnet_direct actions.
+    */
+    action route_vnet(bit<24> dest_vnet_vni,
```
Comment (marian-pritsak) [inline dash-pipeline/bmv2/dash_outbound.p4:34]: Need to split into a separate action vnet_direct

**Review by marian-pritsak — CHANGES_REQUESTED** [review]

File: dash-pipeline/bmv2/dash_outbound.p4 (hunk 1 above)
Comment (mukeshmv) [reply dash-pipeline/bmv2/dash_outbound.p4]: Done pls check.

**Comment by mukeshmv** [comment]
@marian-pritsak @prsunny once this is merged do we need to copy the auto-generated Dash SAI file updates to the OCP SAI repo ?

**Comment by chrispsommers** [comment]
> @marian-pritsak @prsunny once this is merged do we need to copy the auto-generated Dash SAI file updates to the OCP SAI repo ?

@mukeshmv The answer is "no." Currently, all DASH SAI apis are "experimental" and not part of the SAI repo. The experimental APIs are mostly auto-generated from DASH P4 code (some are fixed code), and are ephemeral: they don't get committed to any repo per se. Elsewhere in the DASH project, they will get incorporated into other artifacts like the SAI-thrift server and client (in an upcoming PR). At some point in the future, they can become candidates for formal merging into the "official" SAI APIs.

**Comment by mukeshmv** [comment]
> 

Thanks @chrispsommers. I noticed a  

> > @marian-pritsak @prsunny once this is merged do we need to copy the auto-generated Dash SAI file updates to the OCP SAI repo ?
> 
> @mukeshmv The answer is "no." Currently, all DASH SAI apis are "experimental" and not part of the SAI repo. The experimental APIs are mostly auto-generated from DASH P4 code (some are fixed code), and are ephemeral: they don't get committed to any repo per se. Elsewhere in the DASH project, they will get incorporated into other artifacts like the SAI-thrift server and client (in an upcoming PR). At some point in the future, they can become candidates for formal merging into the "official" SAI APIs.

Thanks @chrispsommers. I noticed a "dash" branch in the OCP SAI repo that has the latest dash experimental files committed. Hence the question.
https://github.com/opencomputeproject/SAI/tree/dash

**Review by marian-pritsak — APPROVED** [review]

**Review by marian-pritsak — CHANGES_REQUESTED** [review]
The API generation yields empty lines:

```
@@ -140,6 +140,7 @@ typedef enum _sai_eni_ether_address_map_entry_attr_t
      *
      * @type sai_object_id_t
      * @flags CREATE_AND_SET
+
      * @objects SAI_OBJECT_TYPE_ENI
      * @allownull true
      * @default SAI_NULL_OBJECT_ID
```

File: dash-pipeline/SAI/sai_api_gen.py [hunk 2]
```diff
@@ -145,6 +147,18 @@ def table_with_counters(program, table_id):
             return 'true'
     return 'false'
 
+def fill_table_attrs(table_attrs, attr_names, action):
```
Comment (marian-pritsak) [inline dash-pipeline/SAI/sai_api_gen.py:150]: The name is misleading. It's not all table attrs, only action parameters.

**Comment by mukeshmv** [comment]
> The API generation yields empty lines:

Fixed

File: dash-pipeline/SAI/sai_api_gen.py (hunk 2 above)
Comment (mukeshmv) [reply dash-pipeline/SAI/sai_api_gen.py]: Changed to fill_table_action_params

**Review by marian-pritsak — CHANGES_REQUESTED** [review]
SAI_OUTBOUND_ROUTING_ENTRY_ATTR_IS_OVERLAY_NH_IP_V6 should not be generated because this flag is already part of SAI_OUTBOUND_ROUTING_ENTRY_ATTR_OVERLAY_NH_IP value sai_ip_address_t.

See https://github.com/Azure/DASH/blob/main/dash-pipeline/SAI/sai_api_gen.py#L187 for reference

**Comment by mukeshmv** [comment]
> SAI_OUTBOUND_ROUTING_ENTRY_ATTR_IS_OVERLAY_NH_IP_V6 should not be generated because this flag is already part of SAI_OUTBOUND_ROUTING_ENTRY_ATTR_OVERLAY_NH_IP value sai_ip_address_t.
> 
> See https://github.com/Azure/DASH/blob/main/dash-pipeline/SAI/sai_api_gen.py#L187 for reference

@marian-pritsak fixed this. Pls check and merge.

**Review by marian-pritsak — APPROVED** [review]

**Comment by marian-pritsak** [comment]
The SAI API change https://github.com/opencomputeproject/SAI/pull/1550

**Pull request merged**